dbms/src/Functions/LeastGreatest.h

```diff
@@ -120,5 +121,363 @@ class FunctionVectorizedLeastGreatest : public IFunction
+    static void process(
+        const TiDB::TiDBCollatorPtr & collator,
+        const ColumnString::Chars_t & a_data,
+        const ColumnString::Offsets & a_offsets,
+        const ColumnString::Chars_t & b_data,
+        const ColumnString::Offsets & b_offsets,
+        ColumnString::Chars_t & c_data,
+        ColumnString::Offsets & c_offsets,
+        size_t i)
+    {
+        size_t a_size = a_offsets[i] - a_offsets[i - 1] - 1;
+        size_t b_size = b_offsets[i] - b_offsets[i - 1] - 1;
+        processImpl(collator, a_size, b_size, &a_data[a_offsets[i - 1]], &b_data[b_offsets[i - 1]], c_data, c_offsets, i);
+    }
+
+    // string_constant
+    static void process(
+        const TiDB::TiDBCollatorPtr & collator,
+        const ColumnString::Chars_t & a_data,
+        const ColumnString::Offsets & a_offsets,
+        const String & b,
+        ColumnString::Chars_t & c_data,
+        ColumnString::Offsets & c_offsets,
+        size_t i)
+    {
+        const auto * b_data = reinterpret_cast<const unsigned char *>(b.data());
+        ColumnString::Offset b_size = b.size();
+        size_t a_size = a_offsets[i] - a_offsets[i - 1] - 1;
+        processImpl(collator, a_size, b_size, &a_data[a_offsets[i - 1]], &b_data[0], c_data, c_offsets, i);
+    }
+
+    // constant_constant
+    static void process(
+        const TiDB::TiDBCollatorPtr & collator,
+        const std::string & a,
+        const std::string & b,
+        std::string & c)
+    {
+        int res = 0;
+        if constexpr (use_collator)
+            res = collator->compare(reinterpret_cast<const char *>(a.data()), a.size(), reinterpret_cast<const char *>(b.data()), b.size());
+        else
+            res = memcmp(a.data(), b.data(), std::min(a.size(), b.size()));
+
+        if constexpr (least)
+        {
+            if (res < 0)
+                c = a;
+            else if (res == 0)
+            {
+                size_t size = std::min(a.size(), b.size());
+                c = b.substr(0, size);
+            }
+            else
+                c = b;
+        }
+        else
+        {
+            if (res < 0)
+                c = b;
+            else if (res == 0)
+            {
+                if (a.size() > b.size())
+                    c = a;
+                else
+                    c = b;
+            }
+            else
+                c = a;
+        }
+    }
+};
+
+template <bool least, bool use_collator>
+struct StringOperationImpl
+{
+    static void NO_INLINE stringVectorStringVector(
+        const ColumnString::Chars_t & a_data,
+        const ColumnString::Offsets & a_offsets,
+        const ColumnString::Chars_t & b_data,
+        const ColumnString::Offsets & b_offsets,
+        const TiDB::TiDBCollatorPtr & collator,
+        ColumnString::Chars_t & c_data,
+        ColumnString::Offsets & c_offsets)
+    {
+        size_t size = a_offsets.size();
+        c_data.resize(std::max(a_data.size(), b_data.size()));
+        if (size == 0)
```

Comment:
size == 0 could be judged after size is assigned, so that we can avoid many useless operation